packages/dataviews/src/dataforms-layouts/data-form-layout.tsx

```diff
@@ -73,12 +103,53 @@ export function DataFormLayout< Item >( {
 					return children( FieldLayout, formField );
 				}
 
+				const errorMessage =
+					( fieldDefinition?.validationSchema?.onTouched &&
+						validation.touchedFields.includes(
+							fieldDefinition.id
+						) ) ||
+					! fieldDefinition?.validationSchema?.onTouched
+						? Object.values(
+								validation.errorMessages[ formField.id ] ?? []
+						  )[ 0 ]
+						: '';
+
 				return (
 					<FieldLayout
 						key={ formField.id }
 						data={ data }
 						field={ formField }
-						onChange={ onChange }
+						errorMessage={ errorMessage }
+						onChange={ ( value ) => {
+							onChange( value );
+							if (
+								! validation.touchedFields?.includes(
+									formField.id
+								)
+							) {
+								validation.setTouchedFields( [
+									...validation.touchedFields,
+									formField.id,
+								] );
+							}
+
+							// Run validation callbacks
+							const errors = Object.entries(
+								fieldDefinition?.validationCallbacks ?? {}
+							).reduce( ( acc, [ key, callback ] ) => {
+								const error = callback( value[ formField.id ] );
+								if ( ! error ) {
+									return acc;
+								}
+
+								return {
+									...acc,
+									[ key ]: error,
+								};
+							}, {} );
+
+							validation.setErrors( formField.id, errors );
```

Comment:
Do we need two callbacks for `onChange` and `setErrors`. There's something that's bothering me here:

 The questions are:

 - Should the validation be performed by `DataForm` or by the consumer of `DataForm` in the `onChange` event handler? (I don't know)
 - If the former, I think we should have `errors` as an extra argument of the `onChange` handler maybe 
 - It might be better to start with the latter though (validation performed on demand)

I feel like we need a good story here of what we want to achieve, it's still not clear to me.

Reply:
I think that the validation should be performed by DataForm, but allow consumers to run their custom logic in their codebase and set errors via the validation object returned by useValidation. Although the extensibility API is still being defined, one of our goals is to allow PHP developers to add or remove fields using only PHP code (from what I understood). If validation is implemented on the consumer side, it will complicate the process for them.

Reply:
> It might be better to start with the latter though (validation performed on demand)

Thinking about this more, I would lean towards this. As I am also not entirely settled with the fact that `onChange` is run first and than the validation/touched states are updated.
While this may work for our current use case, I could see consumers not wanting to update the specific state if there was errors. With this they have no way of knowing.

Passing `errors` as the additional argument for `onChange` would fix this.

Reply:
> Passing errors as the additional argument for onChange would fix this.

I'm going to work on another PR exploring this approach.

Just to clarify, are you imaging an API like this?

```

<DataForm
  data={pattern}
  fields={fields}
  form={form}
  onChange={(newData, errors) => {
    const newErros = validateForm(newData, fields);
    setPattern({
      ...pattern,
      ...newData,
    });
    errors.setErrors(newErrors);
  }}
/>
<Button disabled={!isValidForm}>Save</Button>

```
Few thoughts:
- On-demand validation prevents marking the form as invalid on the first render. For instance, a form with required fields should be initially invalid, while a form without required fields could be valid from the start.
- Shifting validation solely to the consumer side means we can't ensure that a field with a validationSchema will always undergo validation.

Reply:
> Just to clarify, are you imaging an API like this?

Thanks for the example, I was actually thinking of keeping the validation triggered within the `DataForm` component, but that the `onChange` callback includes the errors ( similar to yours above ):
```
onChange={(newData, newErrors) => {
    setPattern({
      ...pattern,
      ...newData,
    });
    errors.setErrors(newErrors);
  }}
```